polkadot/xcm/xcm-executor/src/lib.rs

```diff
@@ -1361,6 +1366,12 @@ impl<Config: config::Config> XcmExecutor<Config> {
 				result
 			},
 			PayFees { asset } => {
+				// If we've already paid for fees, do nothing.
+				if self.already_paid_fees {
+					return Ok(());
+				}
+				// Make sure `PayFees` won't be processed again.
+				self.already_paid_fees = true;
 				// Message was not weighed, there is nothing to pay.
 				if self.message_weight == Weight::zero() {
```

Comment:
not specific to this PR, but caught my eye:

when can `message_weight` be zero? just skipping instruction in that case make sense for `BuyExecution`, but `PayFees` is meant to cover more than just execution costs (e.g. subsequent transport costs or contracts gas costs)

Reply:
It shouldn't be zero actually. Also, as you say, even if weight *was* zero, there's still a benefit in loading the `fees` register for delivery or EVM fees

Reply:
Not clear to me if you want to address this here or in a separate PR/issue or... ?

Reply:
I'll address here